### PR TITLE
Add support for CTC for float64

### PR DESCRIPTION
**Pull request by andrii-uniq**
Previous pull request with this commit #21822

### CODE REVIEW

**Comment by aprimostka** [comment]
Are you sure @rthadur? One more year of waiting from the same reviewer?

**Comment by aprimostka** [comment]
@rthadur, as with #21822, @ebrevdo just ignores this PR

File: tensorflow/core/kernels/ctc_loss_op.cc [hunk 1]
```diff
@@ -19,21 +19,23 @@ limitations under the License.
...
+#include "tensorflow/core/framework/register_types.h"
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/platform/macros.h"
 #include "tensorflow/core/util/ctc/ctc_loss_calculator.h"
 #include "tensorflow/core/util/sparse/sparse_tensor.h"
 
 namespace tensorflow {
 
-typedef Eigen::ThreadPoolDevice CPUDevice;
+// typedef Eigen::ThreadPoolDevice CPUDevice;
```
Comment (ebrevdo) [inline tensorflow/core/kernels/ctc_loss_op.cc]: commented out?

File: tensorflow/core/ops/ctc_ops.cc [hunk 2]
```diff
@@ -25,15 +25,16 @@ using shape_inference::ShapeHandle;
...
     .Input("labels_values: int32")
     .Input("sequence_length: int32")
     .Attr("preprocess_collapse_repeated: bool = false")
     .Attr("ctc_merge_repeated: bool = true")
     .Attr("ignore_longer_outputs_than_inputs: bool = false")
-    .Output("loss: float")
-    .Output("gradient: float")
+    .Output("loss: T")
+    .Output("gradient: T")
+    .Attr("T: {float, double}")
```
Comment (ebrevdo) [inline tensorflow/core/ops/ctc_ops.cc]: make the default float for backwards compatibility

File: tensorflow/core/ops/ctc_ops.cc [hunk 3]
```diff
@@ -90,15 +92,16 @@ REGISTER_OP("CTCGreedyDecoder")
...
     .Input("sequence_length: int32")
     .Attr("beam_width: int >= 1")
     .Attr("top_paths: int >= 1")
     .Attr("merge_repeated: bool = true")
     .Output("decoded_indices: top_paths * int64")
     .Output("decoded_values: top_paths * int64")
     .Output("decoded_shape: top_paths * int64")
-    .Output("log_probability: float")
+    .Output("log_probability: T")
+    .Attr("T: {float, double}")
```
Comment (ebrevdo) [inline tensorflow/core/ops/ctc_ops.cc]: ditto

File: tensorflow/core/util/ctc/BUILD [hunk 4]
```diff
@@ -59,7 +59,8 @@ tf_cc_tests(
     name = "ctc_beam_search_test",
     size = "small",
     srcs = [
-        "ctc_beam_search_test.cc",
+        "ctc_beam_search_test_float.cc",
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/BUILD]: make a single .cc file to test both and use templating within that file.

File: tensorflow/core/util/ctc/ctc_beam_search.h [hunk 5]
```diff
@@ -196,7 +199,7 @@ Status CTCBeamSearchDecoder<CTCBeamState, CTCBeamComparer>::Decode(
     }
 
     Status status =
-        TopPaths(top_n, &beams, &beam_log_probabilities, merge_repeated_);
+        TopPaths(top_n, &beams, &beam_log_probabilities, this->merge_repeated_);
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_beam_search.h]: why?

File: tensorflow/core/util/ctc/ctc_beam_search.h [hunk 6]
```diff
@@ -239,43 +242,43 @@ float CTCBeamSearchDecoder<CTCBeamState, CTCBeamComparer>::GetTopK(
     }
   }
   // Return max value which is in 0th index or blank character logit
-  return std::max((*top_k_logits)[0], input(num_classes_ - 1));
+  return std::max((*top_k_logits)[0], input(this->num_classes_ - 1));
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_beam_search.h]: why?

File: tensorflow/core/util/ctc/ctc_beam_search.h [hunk 7]
```diff
@@ -304,7 +307,7 @@ void CTCBeamSearchDecoder<CTCBeamState, CTCBeamComparer>::Step(
       b->newp.label += raw_input(b->label) - norm_offset;
     }
     // Pblank(l=abc @ t=6) = P(l=abc @ t=5) * P(- @ 6)
-    b->newp.blank = b->oldp.total + raw_input(blank_index_) - norm_offset;
+    b->newp.blank = b->oldp.total + raw_input(this->blank_index_) - norm_offset;
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_beam_search.h]: why?

File: tensorflow/core/util/ctc/ctc_beam_search_test_double.cc [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+typedef tensorflow::ctc::CTCDecoder<double> CTCDecoder;
+
+// The HistoryBeamState is used to keep track of the current candidate and
+// caches the expansion score (needed by the scorer).
+struct HistoryBeamState {
+  double score;
+  std::vector<int> labels;
+};
+
+// DictionaryBeamScorer essentially favors candidates that can still become
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_beam_search_test_double.cc]: put into a separate PR, please.

Comment (ebrevdo) [reply tensorflow/core/util/ctc/ctc_beam_search_test_double.cc]: oh nevermind; just move this into a common test file templated on type float and double.

File: tensorflow/core/util/ctc/ctc_beam_search_test_float.cc [hunk 9]
```diff
@@ -25,8 +25,8 @@ limitations under the License.
 namespace {
 
 typedef std::vector<std::vector<std::vector<float>>> TestData;
-using tensorflow::ctc::CTCBeamSearchDecoder;
-using tensorflow::ctc::CTCDecoder;
+typedef tensorflow::ctc::CTCBeamSearchDecoder<float> CTCBeamSearchDecoder;
+using tensorflow::ctc::CTCDecoder<double> CTCDecoder;
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_beam_search_test_float.cc]: this is test_float.cc but you're using doubles.  again, please make these templated on T and just use macros to build test functions for both float and double.

File: tensorflow/core/util/ctc/ctc_decoder.h [hunk 10]
```diff
@@ -69,25 +70,27 @@ class CTCDecoder {
...
+  Status Decode(const typename CTCDecoder<T>::SequenceLength& seq_len,
+                const std::vector<typename CTCDecoder<T>::Input>& input,
+                std::vector<typename CTCDecoder<T>::Output>* output,
+                typename CTCDecoder<T>::ScoreOutput* scores) override {
+    if (output->empty() || (*output)[0].size() < Decoder::batch_size_) {
       return errors::InvalidArgument(
           "output needs to be of size at least (1, batch_size).");
     }
-    if (scores->rows() < batch_size_ || scores->cols() == 0) {
+    if (scores->rows() < Decoder::batch_size_ || scores->cols() == 0) {
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_decoder.h]: why?

File: tensorflow/core/util/ctc/ctc_decoder.h [hunk 11]
```diff
@@ -69,25 +70,27 @@ class CTCDecoder {
...
       return errors::InvalidArgument(
           "scores needs to be of size at least (batch_size, 1).");
     }
     // For each batch entry, identify the transitions
-    for (int b = 0; b < batch_size_; ++b) {
+    for (int b = 0; b < Decoder::batch_size_; ++b) {
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_decoder.h]: ditto; undo

File: tensorflow/core/util/ctc/ctc_decoder.h [hunk 12]
```diff
@@ -98,8 +101,8 @@ class CTCGreedyDecoder : public CTCDecoder {
         auto row = input[t].row(b);
         int max_class_ix;
         (*scores)(b, 0) += -row.maxCoeff(&max_class_ix);
-        if (max_class_ix != blank_index_ &&
-            !(merge_repeated_ && max_class_ix == prev_class_ix)) {
+        if (max_class_ix != Decoder::blank_index_ &&
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_decoder.h]: ditto; undo

File: tensorflow/core/util/ctc/ctc_loss_calculator.h [hunk 13]
```diff
@@ -357,6 +364,174 @@ Status CTCLossCalculator::PopulateLPrimes(
...
+      // Add in the u + 2, t + 1 term if l_prime(u) != blank or l_prime(u+2).
+      if (u + 2 < U) {
+        const bool matching_labels_merge =
+            ctc_merge_repeated && (l_prime[u] == l_prime[u + 2]);
+        if (l_prime[u] != blank_index_ && !matching_labels_merge) {
+          // Add in u + 2 term.
+          log_beta->coeffRef(u, t) =
+              LogSumExp(log_beta->coeff(u, t),
+                        log_beta->coeff(u + 2, t + 1) +
+                            log(y(l_prime[u + 2], output_delay_ + t + 1)));
```
Comment (ebrevdo) [inline tensorflow/core/util/ctc/ctc_loss_calculator.h]: are you sure this is using the correct log function?  perhaps you should use Eigen::numext::log which is correctly templated (if not using it already).

Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_loss_calculator.h]: This function was moved from ctc_loss_calculator.cc to make it template

File: tensorflow/core/util/ctc/ctc_beam_search.h (hunk 5 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_beam_search.h]: Empirically answer:
/tensorflow/core/util/ctc/ctc_beam_search.h:202:58: error: 'merge_repeated_' was not declared in this scope
         TopPaths(top_n, &beams, &beam_log_probabilities, merge_repeated_);
                                                          ^~~~~~~~~~~~~~~
I don't know the exact standard paragraph.

File: tensorflow/core/kernels/ctc_loss_op.cc (hunk 1 above)
Comment (aprimostka) [reply tensorflow/core/kernels/ctc_loss_op.cc]: resolved in 51cd5877412f9d67f2ec180c8087ae8d7a059da1

File: tensorflow/core/ops/ctc_ops.cc (hunk 2 above)
Comment (aprimostka) [reply tensorflow/core/ops/ctc_ops.cc]: resolved in 51cd5877412f9d67f2ec180c8087ae8d7a059da1

File: tensorflow/core/ops/ctc_ops.cc (hunk 3 above)
Comment (aprimostka) [reply tensorflow/core/ops/ctc_ops.cc]: resolved in 51cd5877412f9d67f2ec180c8087ae8d7a059da1

File: tensorflow/core/util/ctc/BUILD (hunk 4 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/BUILD]: resolved in 51cd5877412f9d67f2ec180c8087ae8d7a059da1

File: tensorflow/core/util/ctc/ctc_beam_search.h (hunk 6 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_beam_search.h]: same as in https://github.com/tensorflow/tensorflow/pull/31164#discussion_r312368512

File: tensorflow/core/util/ctc/ctc_beam_search.h (hunk 7 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_beam_search.h]: same as in https://github.com/tensorflow/tensorflow/pull/31164#discussion_r312368512

File: tensorflow/core/util/ctc/ctc_decoder.h (hunk 10 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_decoder.h]: same as in https://github.com/tensorflow/tensorflow/pull/31164#discussion_r312368512

File: tensorflow/core/util/ctc/ctc_decoder.h (hunk 11 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_decoder.h]: same as in https://github.com/tensorflow/tensorflow/pull/31164#discussion_r312368512

File: tensorflow/core/util/ctc/ctc_decoder.h (hunk 12 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_decoder.h]: same as in https://github.com/tensorflow/tensorflow/pull/31164#discussion_r312368512

File: tensorflow/core/util/ctc/ctc_beam_search_test_double.cc (hunk 8 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_beam_search_test_double.cc]: resolved in 51cd5877412f9d67f2ec180c8087ae8d7a059da1

File: tensorflow/core/util/ctc/ctc_beam_search_test_float.cc (hunk 9 above)
Comment (aprimostka) [reply tensorflow/core/util/ctc/ctc_beam_search_test_float.cc]: resolved in 51cd5877412f9d67f2ec180c8087ae8d7a059da1

**Review by ebrevdo — APPROVED** [review]
Assuming all tests are green; let's submit.  We'll then test internally.

**Comment by aprimostka** [comment]
Should I do something to fix the error with the failed build? Because I have no idea what to do with this specific error.

**Comment by ebrevdo** [comment]
I will try and fix it upon merge.
